<a href="https://colab.research.google.com/github/molecular-mar/molecular-mar.github.io/blob/master/Sesion5_1_PAQ24I.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Principios de programación 4
---

## Ciclos
---

### Ciclo `for`


#### Sintaxis

Una herramienta básica en programación es el concepto de *ciclo*: una instrucción que señala que otro grupo de instrucciones deben de realizarse de forma repetida. Uno de estos ciclos se conoce como **for**, y su sintaxis es la siguiente:
```python
for variable in iterable:
  instruccion_1
  instruccion_2
  ...
  instruccion_n
```
Revisemos a detalle esta sintaxis. Comenzamos por la instrucción `for`(nota el color azul del texto), que indica el tipo de ciclo a usar. Luego, después de un espacio indicamos una `variable`. Después de otro espacio, tenemos la instrucción `in`, y luego de otro espacio encontramos un `iterable`, junto con el símbolo de dos puntos (`:`).

Un *iterable* en Python es un tipo de dato sobre el cuál podemos realizar un ciclo, recuperando uno de los valores que lo componen cada la vez. Sin entrar a mucho detalle, en general podemos pensar que los iterables son como las colecciones de datos, como las listas, las tuplas y los strings abordadas en la sesión anterior.


#### Identación

Antes de experimentar con el uso de `for`, un **punto importante** a considerar. Es importante poder distinguir entre instrucciones asociadas al ciclo e instrucciones no asociadas. En Python es muy importante la *identación*, el espacio que hay al inicio de una linea de código. Las líneas contiguas de código con una misma identación están en un mismo *bloque*, lo que afecta su comportamiento. Por lo tanto, este espacio al inicio no debe ser tomado a la ligera, sino utilizado a conciencia.

Hay dos convenciones de uso frecuente: usar 2 espacios o 4 espacios. Para este curso puedes utilizar la que prefieras, pero debes ser consistente.

❓Copia la lista mm_elementos de la Sesión anterior en una cc. En una celda separada o debajo de la definición, escribe este fragmento de código:
```python
for masa in mm_elementos:
    print(masa)
    print('Probando el ciclo for')
```
Observa que ocurre y escribe con tus propias palabras qué es lo que hace el ciclo for (puedes partir de la traducción literal al español de la instrucción).

❓ Realiza los siguientes ajustes sobre el `for` anterior, duplicando cada vez la celda de código que recién creamos. En cada caso, documenta tu observación:
* Comenta la linea de `print(masa)`.
* Retira la identación de la linea `print('Probando el ciclo for')`.


❓La siguiente celda contiene una lista de concentraciones en g/mL. Asumiendo que el soluto es acetato de sodio (CH$_3$COONa), calcula la molaridad (mol/L) dentro de un ciclo `for`, imprimiendo la concentración obtenida con un mensaje adecuado.

In [ ]:
concentraciones = [0.05, 0.25, 1/3, 2.5, 2.E-2, 0.60, 4.25]

#### La función `range`

En el ejemplo anterior usamos una lista como iterable en el ciclo `for`. Cuando sabemos cuántas veces queremos repetir un ciclo, y no queremos *iterar* sobre una lista, podemos utilizar la función `range`, la cuál genera un iterable que cumple con cierto *rango* de valores. Realizemos algunas pruebas para entender su funcionamiento.

❓Escribe un ciclo `for`, usando como iterable `range(5)`. Escribe al menos una instrucción `print` dentro del ciclo `for` para observar los valores de la variable asociada al ciclo. Documenta tu observación.

La función `range` no genera una lista, pero si algo similar. Podemos convertir el tipo de dato que genera `range` en tipo lista, usando la función `list()` (recuerda las funciones `int()`, `float()` y `str()`).
Ejecuta como ejemplo la siguiente celda:

In [ ]:
contando = list(range(10))
print(contando)

Hay algunas formas de cambiar el comportamiento de `range`. Por default siempre se comienza en 0, pero si en lugar de un parámetro indicamos dos, el primero se utilizará como punto de partida:

In [ ]:
list(range(5,10))

Podemos incluir un tercer parámetro que modifique el tamaño de paso:

In [ ]:
list(range(2,10,2))

❓ Genera una lista que contenga los números impares entre 0 y 10.


#### `for` corto

Hay una forma breve de usar el ciclo `for` cuando se desea crear una lista. A esta forma también se le conoce como *comprensión de lista*:

```python
lista = [elemento_de_la_nueva_lista for var in iterable]
```

El siguiente ejemplo convierte una lista de concentraciones de $H^+$ en valores de pH:

In [3]:
import math
concentraciones_Hmas = [1E3, 1E6, 1E8, 1E13]
pH_calculados = [-math.log10(con) for con in concentraciones_Hmas]
print(pH_calculados)

[-3.0, -6.0, -8.0, -13.0]


#### Ciclo `while`
Otra forma de realizar de forma repetitiva una serie de tareas es usando el ciclo `while`. Para utilizarlo, se fija una condición, y las tareas asociadas al `while` se repiten mientras dicha condición se cumpla. La sintaxis es:

```python
while condición:
  ...#Tareas a repetir
  ...
```
Veremos a mayor detalle como establecer condiciones en la sesión 6.1. En general, se recomienda usar `for` por encima de `while`, ya que usando `while` es probable caer en un problema de *ciclo infinito*, en el cuál la condición siempre se cumple.

❓Ejecuta la siguiente celda de código, que contiene un ejemplo de ciclo infinito.

In [ ]:
vol_naoh = 0 # volumen en mL
while vol_naoh < 100:
  print("Falta agregar NaOH")

#### Algunas tareas comúnes

Existen algunos procedimientos generales que es común realizar dentro de ciclos. Uno de ellos es construir nuevas listas. Para ello es importante crear antes del ciclo la lista que vamos a modificar. Si nuestra lista va a ser creada desde cero, tendremos que definirla como una *lista vacia*:
```python
lista_nueva = [] # Lista vacia
for dato in iterable:
    lista_nueva.append() # Adentro de append() lo que deseamos agregar
```

Otra tarea común es *acumular* valores en una variable. Veamos un ejemplo para aclararlo. Supongamos que estamos realizando una titulación, añadiendo 1.03 mL de ácido en cada gota (estamos usando una bureta automatizada que agrega consistentemente este valor). En un lapso de 19 gotas, queremos saber cuál es el volumen añadido de ácido.

Una posible forma usando un ciclo `for` sería:




In [4]:
var = 3**3 +2
print(var)
var = var + 1
print(var)

29
30


In [6]:
vol_acido = 0 # Definimos un valor inicial
lista_volumenes = []
for paso in range(19):
    vol_acido = vol_acido + 1.03 # Acumulamos el valor
    lista_volumenes.append(vol_acido)
    print(vol_acido)
print(lista_volumenes)

1.03
2.06
3.09
4.12
5.15
6.180000000000001
7.210000000000001
8.24
9.27
10.299999999999999
11.329999999999998
12.359999999999998
13.389999999999997
14.419999999999996
15.449999999999996
16.479999999999997
17.509999999999998
18.54
19.57
[1.03, 2.06, 3.09, 4.12, 5.15, 6.180000000000001, 7.210000000000001, 8.24, 9.27, 10.299999999999999, 11.329999999999998, 12.359999999999998, 13.389999999999997, 14.419999999999996, 15.449999999999996, 16.479999999999997, 17.509999999999998, 18.54, 19.57]


❓Duplica la celda anterior, y añade en la copia un `print` de la variable `vol_acido` dentro del ciclo.

Lo primero que debemos hacer para acumular un valor es definir la variable en donde haremos la acumulación. Luego, dentro del ciclo, debemos definir una operación como la mostrada arriba.

Cuando realizamos una *asignación de variable*, que ocurre cuando usamos el operador `=` (se le llama *de asignación* en programación, no igual), internamente lo primero que se hace es evaluar del lado derecho de `=`, y luego de esto asignar ese valor a la variable. Por ello es posible utilizar a la variable para definir a la variable en este contexto.

El operador `+=` funciona igual que la operación de acumulación mostrada, pero sin la necesidad de repetir el nombre de la variable:
```python
variable = 10
variable += 5 # agregar 5 al valor de variable
```

❓En una copia de la celda de la titulación, sustituye la operación de acumulación por su equivalente usando `+=`.

In [7]:
variable = 10
print(variable)
variable += 5
print(variable)

10
15


#### Ciclos anidados

Es posible crear un ciclo dentro de otro ciclo. Basta con utilizar una identación adecuada. Supongamos que titularemos una solución de 20 mL, añadiendo directamente 5 mL de un aditivo cada que la bureta ha añadido 5 gotas, y queremos saber el volumen total de la solución luego de repetir 3 veces este procedimiento. El siguiente ciclo ilustra este proceso:

In [8]:
vol_total = 20 #mL
for aditivo in range(3):
    print("El ciclo de aditivo es:", aditivo)
    for gota in range(5):
      print("El ciclo de gota es:", gota)
      vol_total += 1.03
    print("Fin de los ciclo de gota. Num:", gota)
    vol_total += 5
print("Fin de los ciclo de aditivo. Num:", aditivo)
print(vol_total)

El ciclo de aditivo es: 0
El ciclo de gota es: 0
El ciclo de gota es: 1
El ciclo de gota es: 2
El ciclo de gota es: 3
El ciclo de gota es: 4
Fin de los ciclo de gota. Num: 4
El ciclo de aditivo es: 1
El ciclo de gota es: 0
El ciclo de gota es: 1
El ciclo de gota es: 2
El ciclo de gota es: 3
El ciclo de gota es: 4
Fin de los ciclo de gota. Num: 4
El ciclo de aditivo es: 2
El ciclo de gota es: 0
El ciclo de gota es: 1
El ciclo de gota es: 2
El ciclo de gota es: 3
El ciclo de gota es: 4
Fin de los ciclo de gota. Num: 4
Fin de los ciclo de aditivo. Num: 2
50.45000000000002


❓Usando ciclos anidados, imprime los números cuánticos $n$ y $l$ permitidos hasta $n=3$. Recuerda que $l$ puede tomar valores entre 0 y $n-1$.

In [10]:
#como imprimir un triangulo

for altura in range(10):
 for largo in range(altura):
  print("*",end='')
 print("")


*
**
***
****
*****
******
*******
********
*********


In [13]:
for n in range(4):
  print("El valor de n es:", n)
  for l in range(n+1):
    print("l=",l)

El valor de n es: 0
l= 0
El valor de n es: 1
l= 0
l= 1
El valor de n es: 2
l= 0
l= 1
l= 2
El valor de n es: 3
l= 0
l= 1
l= 2
l= 3


In [14]:
for n in range(1,4):
  print("n=", n)
  for l in range(n):
    print("l=", l)

n= 1
l= 0
n= 2
l= 0
l= 1
n= 3
l= 0
l= 1
l= 2


In [21]:
print("uamito cena",end='/')
print("sandwich")

uamito cena/sandwich


In [24]:
matriz = []
for i in range(3):
  renglon = []
  for j in range(3):
    renglon.append((i*3)+j)
  matriz.append(renglon)
print(matriz)

[[0, 1, 2], [3, 4, 5], [6, 7, 8]]


In [27]:
print(len(matriz[0]))
for i in range(len(matriz)):
  print(matriz[i])

3
[0, 1, 2]
[3, 4, 5]
[6, 7, 8]


In [29]:
for i in range(len(matriz)):
  for j in range(len(matriz[i])):
    print(matriz[i][j],end=' ')
  print('')

0 1 2 
3 4 5 
6 7 8 


In [31]:
for renglon in matriz:
  print(renglon)

for renglon in matriz:
  for elemento in renglon:
    print(elemento,end='')
  print('')

[0, 1, 2]
[3, 4, 5]
[6, 7, 8]
012
345
678


## Tarea

Resuelve los siguientes problemas en un nuevo notebook.

* Necesitamos realizar una serie de disoluciones con concentraciones diferentes para un experimento. Cuentas con un pequeño robot que puede realizarlas por ti, pero necesitas crear un programa para que funcione. Crea un programa que calcule los gramos necesarios de una lista de compuestos para elaborar una serie de soluciones. Los compuestos son: hidróxido de sodio, triclorometano, acetato de sodio y dimetilformamida. Las concentraciones que deseas son 0.001M, 0.05M, 0.2 M, 0.66 M y 2.1 M. Debes hacer uso de `for`.

* Genera listas para las siguientes funciones. Incluye una gráfica en cada caso (ve la Sesión 5.2):
 * $x^2 + 1/x$. $x\in [-5,5]$
 * $e^{-x^2}$. $x\in [0,20]$
 * $-R_H(\frac{1}{n^2})$. $n\in [1,10]$. $R_H = 1.09677576\times 10^7$

* Usando ciclos `for`, realiza las siguientes operaciones:

 1. Crea las matrices (usando listas anidadas):

$$A= \begin{bmatrix}
1 & 3 & 5\\
3 & 5 & 7\\
5 & 7 & 9\\
\end{bmatrix}$$

$$B= \begin{bmatrix}
2 & 4 & 8\\
4 & 8 & 10\\
8 & 10 & 12\\
\end{bmatrix}$$

 2. Calcula la traza de cada matrix. La traza se define como la suma de los elementos diagonales ($\sum_i A_{ii}$)
 3. Calcula la matriz que resulta de la suma de ambas matrices.
